ERC: Account Abstraction via Entry Point Contract specification (#4337)
ethereum/EIPs: EIPS/eip-4337.md

---
eip: 4337

> title: Account Abstraction via Entry Point Contract specification

Author: ```suggestion
title: Account Abstraction via Entry Point Contract specification
description: An account abstraction proposal which completely avoids consensus-layer protocol changes, instead relying on higher-layer infrastructure.
```

Reviewer: 37954f91ddcec725a03c9e7fa6dabf782c36a9af

author: Vitalik Buterin, Yoav Weiss, Kristof Gazso, Namra Patel, Dror Tirosh
discussions-to: https://ethereum-magicians.org/t/erc-4337-account-abstraction-via-entry-point-contract-specification/7160
type: Standards Track
category: ERC
status: Draft
created: 2021-09-29
---

> ## Simple Summary

Reviewer 2: This section has been deprecated in favor of the `description` field in the preamble. You can see how in should be structured in the [`eip-template.md`](https://github.com/ethereum/EIPs/blob/master/eip-template.md).

An account abstraction proposal which completely avoids the need for consensus-layer protocol changes, instead relying on a separate mempool of `UserOperation` objects and miners either running custom code or connecting to a bundle marketplace.

## Motivation

> See also: https://ethereum-magicians.org/t/implementing-account-abstraction-as-part-of-eth1-x/4020 and the links therein for historical work and motivation, and [EIP 2938](https://eips.ethereum.org/EIPS/eip-2938) for a consensus layer proposal for implementing the same goal.

Reviewer 2: ```suggestion
See also ["Implementing Account Abstraction as Part of Eth 1.x"](https://ethereum-magicians.org/t/implementing-account-abstraction-as-part-of-eth1-x/4020) and the links therein for historical work and motivation, and [EIP-2938](https://eips.ethereum.org/EIPS/eip-2938) for a consensus layer proposal for implementing the same goal.
```

This proposal takes a different approach, avoiding any adjustments to the consensus layer. It seeks to achieve the following goals:

> * **Achieve the key goal of account abstraction**: allow users to use smart contract wallets containing arbitrary verification logic instead of EOAs as their primary account. Completely remove any need at all for users to also have EOAs (as status quo SC wallets and EIP 3074 both require)

Reviewer 2: ```suggestion
* **Achieve the key goal of account abstraction**: allow users to use smart contract wallets containing arbitrary verification logic instead of EOAs as their primary account. Completely remove any need at all for users to also have EOAs (as status quo SC wallets and EIP-3074 both require)
```

* **Decentralization**
    * Allow any bundler (think: miner) to participate in the process of including account-abstracted user operations
    * Work with all activity happening over a public mempool; users do not need to know the direct communication addresses (eg. IP, onion) of any specific actors
    * Avoid trust assumptions on bundlers
* **Do not require any Ethereum consensus changes**: Ethereum consensus layer development is focusing on the merge and later on scalability-oriented features, and there may not be any opportunity for further protocol changes for a long time. Hence, to increase the chance of faster adoption, this proposal avoids Ethereum consensus changes.
* **Try to support other use cases**
    * Privacy-preserving applications
    * Atomic multi-operations (similar goal to EIP 3074)

>     * Pay tx fees with ERC20 tokens, allow developers to pay fees for their users, and [EIP 3074](https://eips.ethereum.org/EIPS/eip-3074)-like **sponsored transaction** use cases more generally

Reviewer 2: ```suggestion
    * Pay tx fees with ERC-20 tokens, allow developers to pay fees for their users, and [EIP-3074](https://eips.ethereum.org/EIPS/eip-3074)-like **sponsored transaction** use cases more generally
```

* **Create the wallet if it does not yet exist**, using the initcode provided in the `UserOperation` (if the wallet does not exist _and_ the initcode is empty, the call must fail)
* **Call `validateUserOp` on the wallet**, passing in the `UserOperation` and the required fee. The wallet should verify the operation's signature, and pay the fee if the wallet considers the operation valid. If any `validateUserOp` call fails, `handleOps` must skip execution of at least that operation, and may revert entirely.

In the execution loop, the `handleOps` call must perform the following steps for each `UserOperation`:

* **Call the wallet with the `UserOperation`'s calldata**. It's up to the wallet to choose how to parse the calldata; an expected worlflow is for the wallet to have an `execute` function that parses the remaining calldata as a series of one or more calls that the wallet should make.
* **Refund unused gas fees** to the wallet

> ![](https://storage.googleapis.com/ethereum-hackmd/upload_4b1c784e1563bb6ddf26886e86e59177.png)

Reviewer 2: This file should be added to the PR. See EIP-1 for the exact directory structure, but I believe it should be stored in `/assets/eip-4337/image.png`.

Author: ```suggestion
![](/assets/eip-4337/image1.png)
```

Before accepting a `UserOperation`, bundlers must use an RPC method to locally simulate calling the `simulateWalletValidation` function of the entry point, to verify that the signature is correct and the operation actually pays fees; see the [Simulation section below](#Simulation) for details.

### Extension: paymasters

> We extend the entry point logic to support **paymasters** that can sponsor transactions for other users. This feature can be used to allow application developers to subsidize fees for their users, allow users to pay fees with ERC20 tokens and many other use cases. When the paymaster is not equal to the zero address, the entry point implements a different flow:

Reviewer 2: ```suggestion
We extend the entry point logic to support **paymasters** that can sponsor transactions for other users. This feature can be used to allow application developers to subsidize fees for their users, allow users to pay fees with ERC-20 tokens and many other use cases. When the paymaster is not equal to the zero address, the entry point implements a different flow:
```

> ![](https://i.imgur.com/fvdcIH5.png)

Reviewer 2: Same with this image.

Author: ```suggestion
![](/assets/eip-4337/image2.png)
```

The entry point-based approach allows for a clean separation between verification and execution, and keeps wallets' logic simple. The alternative would be to require wallets to follow a template where they first self-call to verify and then self-call to execute (so that the execution is sandboxed and cannot cause the fee payment to revert); template-based approaches were rejected due to being harder to implement, as existing code compilation and verification tooling is not designed around template verification.

### Paymasters

Paymasters facilitate transaction sponsorship, allowing third-party-designed mechanisms to pay for transactions. Many of these mechanisms _could_ be done by having the paymaster wrap a `UserOperation` with their own, but there are some important fundamental limitations to that approach:

> * No possibility for "passive" paymasters (eg. that accept fees in some ERC20 token at an exchange rate pulled from an on-chain DEX)

Reviewer 2: ```suggestion
* No possibility for "passive" paymasters (eg. that accept fees in some ERC-20 token at an exchange rate pulled from an on-chain DEX)
```

* Paymasters run the risk of getting griefed, as users could send ops that appear to pay the paymaster but then change their behavior after a block

> The paymaster scheme allows a contract to passively pay on users' behalf under arbitrary conditions. It even allows ERC20 token paymasters to secure a guarantee that they would only need to pay if the user pays them: the paymaster contract can check that there is sufficient approved ERC20 balance in the `validatePaymasterUserOp` method, and then extract it with `transferFrom` in the `postOp` call; if the op itself transfers out or de-approves too much of the ERC20s, the inner `postOp` will fail and revert the execution and the outer `postOp` can extract payment (note that because of storage access restrictions the ERC20 would need to be a wrapper defined within the paymaster itself).

Reviewer 2: ```suggestion
The paymaster scheme allows a contract to passively pay on users' behalf under arbitrary conditions. It even allows ERC-20 token paymasters to secure a guarantee that they would only need to pay if the user pays them: the paymaster contract can check that there is sufficient approved ERC-20 balance in the `validatePaymasterUserOp` method, and then extract it with `transferFrom` in the `postOp` call; if the op itself transfers out or de-approves too much of the ERC-20s, the inner `postOp` will fail and revert the execution and the outer `postOp` can extract payment (note that because of storage access restrictions the ERC-20 would need to be a wrapper defined within the paymaster itself).
```

This is accomplished by having the entry point itself create wallets using CREATE2. The `UserOperation` struct has an `initCode` field; this field would be empty for all operations by a given wallet after the first, but the first operation would fill in the `initCode`. The entry point creates the wallet, and then performs the operation. The user can compute the address of their wallet by locally running the [EIP 1014](https://eips.ethereum.org/EIPS/eip-1014) CREATE2 address formula. The salt used is the `nonce` of the `UserOperation`.

### Entry point upgrading

Wallets are encouraged to be DELEGATECALL forwarding contracts for gas efficiency and to allow wallet upgradability. The wallet code is expected to hard-code the entry point into their code for gas efficiency. If a new entry point is introduced, whether to add new functionality, improve gas efficiency, or fix a critical security bug, users can self-call to replace their wallet's code address with a new code address containing code that points to a new entry point. During an upgrade process, it's expected that two mempools will run in parallel.

> ## Entry point contract implementation

Reviewer 2: ```suggestion
## Reference Implementation
```

---
eip: 4337
title: Account Abstraction via Entry Point Contract specification
description: An account abstraction proposal which completely avoids consensus-layer protocol changes, instead relying on higher-layer infrastructure.

> author: Vitalik Buterin, Yoav Weiss, Kristof Gazso, Namra Patel, Dror Tirosh

Author: ```suggestion
author: Vitalik Buterin (@vbuterin), Yoav Weiss (@yoavw), Kristof Gazso (@kristofgazso), Namra Patel (@namrapatel), Dror Tirosh (@drortirosh) 
```

---
eip: 4337

> title: Account Abstraction via Entry Point Contract specification

Reviewer 3: 3CHevK8twiZnAMDt19XWcPrZ1PBUd5S5ng

description: An account abstraction proposal which completely avoids consensus-layer protocol changes, instead relying on higher-layer infrastructure.

> author: Vitalik Buterin (@vbuterin), Yoav Weiss (@yoavw), Kristof Gazso (@kristofgazso), Namra Patel (@namrapatel), Dror Tirosh (@drortirosh) 

Reviewer 4: ```suggestion
 author: Vitalik Buterin (@vbuterin), Yoav Weiss (@yoavw), Kristof Gazso (@kristofgazso), Namra Patel (@namrapatel), Dror Tirosh (@drortirosh)
```

>  author: Vitalik Buterin (@vbuterin), Yoav Weiss (@yoavw), Kristof Gazso (@kristofgazso), Namra Patel (@namrapatel), Dror Tirosh (@drortirosh)

Reviewer 4: ```suggestion
author: Vitalik Buterin (@vbuterin), Yoav Weiss (@yoavw), Kristof Gazso (@kristofgazso), Namra Patel (@namrapatel), Dror Tirosh (@drortirosh)
```

discussions-to: https://ethereum-magicians.org/t/erc-4337-account-abstraction-via-entry-point-contract-specification/7160
type: Standards Track
category: ERC
status: Draft
created: 2021-09-29
---

## Abstract

An account abstraction proposal which completely avoids the need for consensus-layer protocol changes. Instead of adding new protocol features and changing the bottom-layer transaction type, this proposal instead introduces a higher-layer pseudo-transaction object called a `UserOperation`. Users send `UserOperation` objects into a separate mempool. A special class of actor called bundlers (either miners, or users that can send transactions to miners through a bundle marketplace) package up a set of these objects into a transaction making a `handleOps` call to a special contract, and that transaction then gets included in a block.

## Motivation

> See also ["Implementing Account Abstraction as Part of Eth 1.x"](https://ethereum-magicians.org/t/implementing-account-abstraction-as-part-of-eth1-x/4020) and the links therein for historical work and motivation, and [EIP-2938](./eip-2938.md) for a consensus layer proposal for implementing the same goal.

Reviewer 3: 3CHevK8twiZnAMDt19XWcPrZ1PBUd5S5ng

* **Create the wallet if it does not yet exist**, using the initcode provided in the `UserOperation` (if the wallet does not exist _and_ the initcode is empty, the call must fail)
* **Call `validateUserOp` on the wallet**, passing in the `UserOperation` and the required fee. The wallet should verify the operation's signature, and pay the fee if the wallet considers the operation valid. If any `validateUserOp` call fails, `handleOps` must skip execution of at least that operation, and may revert entirely.

In the execution loop, the `handleOps` call must perform the following steps for each `UserOperation`:

* **Call the wallet with the `UserOperation`'s calldata**. It's up to the wallet to choose how to parse the calldata; an expected worlflow is for the wallet to have an `execute` function that parses the remaining calldata as a series of one or more calls that the wallet should make.
* **Refund unused gas fees** to the wallet

![](../assets/eip-4337/image1.png)

> Before accepting a `UserOperation`, bundlers must use an RPC method to locally simulate calling the `simulateWalletValidation` function of the entry point, to verify that the signature is correct and the operation actually pays fees; see the [Simulation section below](#Simulation) for details.

Reviewer 2: ```suggestion
Before accepting a `UserOperation`, bundlers must use an RPC method to locally simulate calling the `simulateWalletValidation` function of the entry point, to verify that the signature is correct and the operation actually pays fees; see the [Simulation section below](#simulation) for details.
```

![](../assets/eip-4337/image2.png)

During the verification loop, in addition to calling `validateUserOp`, the `handleOps` execution also must check that the paymaster has enough ETH staked with the entry point to pay for the operation, and then call `validatePaymasterUserOp` on the paymaster to verify that the paymaster is willing to pay for the operation. Additionally, the `validateUserOp` must be called with a `requiredPrefund` of 0 to reflect that it's the paymaster, and not the wallet, that's paying the fees.

During the execution loop, the `handleOps` execution must call `postOp` on the paymaster after making the main execution call. It must guarantee the execution of `postOp`, by making the main execution inside an inner call context, and if the inner call context reverts attempting to call `postOp` again in an outer call context.

> Maliciously crafted paymasters _can_ DoS the system. To prevent this, we use a paymaster reputation system; see the [reputation, throttling and banning section](#Reputation-scoring-and-throttling%2Fbanning-for-paymasters) for details.

Reviewer 2: ```suggestion
Maliciously crafted paymasters _can_ DoS the system. To prevent this, we use a paymaster reputation system; see the [reputation, throttling and banning section](#reputation-scoring-and-throttlingbanning-for-paymasters) for details.
```
